### PR TITLE
Add GDScript file type

### PR DESCRIPTION
**Pull request by Davidy22**
[Godot](https://github.com/godotengine/godot) has been doing a tour de force lately as an open source alternative to Unity. Add support for the growing project's GDScript.

Heavily borrowed from python due to GDScript's syntax being heavily borrowed from python. To note, the language is undergoing some breaking changes that will take effect when godot 4 releases, this already should account for syntactical changes and I've already classified some number of currently-keywords as builtins even though godot 4 isn't released yet.

A little test file that I used to make sure I didn't completely break it while making the changes for the things that are different between python and gdscript:
image

And the prerequisite upstream pull requests:
https://github.com/ScintillaOrg/lexilla/pull/41 - Merged
https://github.com/universal-ctags/ctags/pull/3194, superceded by https://github.com/universal-ctags/ctags/pull/3199 - Merged

### CODE REVIEW

**Comment by techee** [comment]
Just had a brief look at the PR and it looks good to me in principle, just the scintilla part and the uctags part should be merged upstrem before applying in Geany (with the corresponding changes that the upstram projects request). Then, also drop the `geany_` prefix from the parser file name - it just indicates those parsers where we don't use the upstream version yet.

**Comment by Davidy22** [comment]
Will update when upstream pulls, the code in this pull request is fairly out of date currently, haven't copied changes into this for a minute

**Comment by Davidy22** [comment]
Alright, all upstream sources have merged, copied everything back into here. There is one general function added to ctags entry.c that was created for the GDScript ctag that is probably not in a ctags release yet because it's brand new but I put it in the same place that it is in the upstream. Also did the rename, although I notice every other ctag parser is prepended with geany_

**Comment by elextr** [comment]
Better wait for #2990 and #2991 in case they cause conflicts.

**Comment by Davidy22** [comment]
Alright. I'll use the branch in the meantime, see if it crashes me out or anything

**Comment by Davidy22** [comment]
Alright, got the notification for the merge, went and fixed the one merge conflict

**Comment by Davidy22** [comment]
Updated and resolved merge conflicts now that preceding PRs have been merged

**Comment by techee** [comment]
@Davidy22 Could you make the patch on top of current master? There was a problem with the update to new Scintilla that caused crashes and your branch crashed because of that too.

I just had a brief look at the pull request. Could you add some unit test that generates all the kinds supported by the parser? Also, I expect the parser generates full scope information - in such case it should be mentioned in the top part of `tm_parser_has_full_context()`.

**Comment by Davidy22** [comment]
Oh, is there new stuff? Rebased again, it did pass CI before rebase though, does something there need to be added to tests as well?

There's a couple of test cases in the upstream ctags that were added as part of the upstream addition, I can copy some of them into test/ctags. Added some lines in lines in tm_parser as well.

**Comment by techee** [comment]
Thanks. For the unit tests to work though, more work is needed than just copying them to the test directory. You need to:
1. Update `tests/ctags/Makefile.am`
2. Create empty `tests/ctags/<fname>.tags` where `<fname>` is the filename of each of the tests
3. Re-run `configure`
4. Run `make check` which will fail but which will create `<fname>.log` files inside the `tests/ctags` directory
5. Inside `tests/ctags` run `patch -p0 <fname>.log` for each of the files
6. Re-run `make check` to make sure all tests pass
7. Commit the `<fname>.tags` files and the `Makefile.am` change

**Comment by techee** [comment]
I'd also suggest editing `tm_parser_enable_role()` and adding something like:
```		
case TM_PARSER_GDSCRIPT:
	return kind == 'c' ? FALSE : TRUE;
```
otherwise both base classes and real classes are shown in the symbol tree (check e.g. `gdscript-inner-class.gd` where you don't want to see the `Reference` tags in the sidebar).

**Comment by Davidy22** [comment]
Alright done

File: tests/ctags/Makefile.am [hunk 1]
```diff
@@ -147,6 +147,9 @@ test_sources = \
 	cxx11enum.cpp					\
 	cxx11-final.cpp					\
 	cxx11-noexcept.cpp				\
+	gdscript-inner-class.gd			\
+	gdscript-modifiers.gd			\
+	gdscript-no-implicit-class.gd	\
```
Comment (techee) [inline tests/ctags/Makefile.am:152]: Just a nitpicky comment: better to sort the files alphabetically.

**Comment by techee** [comment]
The PR looks good to me. I did test it a bit and everything seems to work and while I'm not a Scintilla highlighting expert, the scintilla part looks good too. I believe the PR could be merged.

@elextr Whet do you think?

**Comment by elextr** [comment]
Not at my dev machine for a couple of weeks so I can't test, so I did a totally unreliable scan of the changes.

I notice that there are changes to files in `ctags/main`.  I thought all those were just copied verbatim from uctags, so any changes there will be overwritten next time it is upgraded?

File: tests/ctags/Makefile.am (hunk 1 above)
Comment (Davidy22) [reply tests/ctags/Makefile.am:152]: I think it goes hijklmn, so these are already in alphabetical order?

Comment (Davidy22) [reply tests/ctags/Makefile.am:152]: Oh wait somehow I inserted these in a different place on the list than I thought I did, adjusting

**Comment by Davidy22** [comment]
The changes in ctags/main are changes that were made in uctags very shortly after the point that the latest ctags update was done from. You should be able to find those functions in upstream, and they should still be there after the next upgrade.

**Comment by techee** [comment]
> The changes in ctags/main are changes that were made in uctags very shortly after the point that the latest ctags update was done from. You should be able to find those functions in upstream, and they should still be there after the next upgrade.

Yeah, I noticed those too and these should be fine.

**Comment by elextr** [comment]
Ok, otherwise I didn't spot anything S/B fine

**Comment by techee** [comment]
@elextr OK if I merge it?

Sorry if I sound too impatient but I'd like to speed-up the ctags parser merging process a little. I really appreciate your desire for Geany stability and I'm the last who wants to introduce some source of instability to Geany but there are many more parsers to be merged (for which I'd like to create pull requests after improving the symbol tree category mapping a little).

I actually think that keeping the code in separate pull requests is bad for Geany release stability because people don't test these regularly and more eyes see more things in master (like in the case of Scintilla where you alone didn't experience the crash).

It's not that I think that every pull request should be merged immediately (I for instance tend to have a 2 week self-correction phase) but if it's something that's generally agreed on thing like that we want to use the upstream parsers or Scintilla and that doesn't require some extensive review (when merging something like a new implementation of a parser, it's basically unreviewable), I'd suggest merging it soon instead of "soon" :-).

**Comment by elextr** [comment]
> OK if I merge it?

thats what my "Ok" above was meant to convey, 4 hours ago, isn't it merged yet, whats the delay :grin:

**Comment by techee** [comment]
One more note - once all the new parsers and related tag manager changes are in, I'd like to do a very extensive testing of all the filetypes and check if things display correctly in the symbol tree and if everything works as expected. But it's a loss of time right now when not everything is in the tag manager yet.

**Comment by techee** [comment]
> thats what my "Ok" above was meant to convey, 4 hours ago, isn't it merged yet, whats the delay 😁

:-)

**Comment by techee** [comment]
@Davidy22 Thanks!

**Comment by Davidy22** [comment]
woo

**Comment by elextr** [comment]
@Davidy22 and thanks for your patience and perseverance

@techee I don't disagree with most of what you said [above](https://github.com/geany/geany/pull/3012#issuecomment-1013670170), except that an unstable Geany means anybody who tests has to not use as their production tool, and well, that means it won't get much more testing than individual PRs will.  Its a damned if we do, damned if we don't situation, sigh.

**Comment by techee** [comment]
> @techee I don't disagree with most of what you said above, except that an unstable Geany means anybody who tests has to not use it as their production tool, and well, that means it won't get much more testing than individual PRs will. Its a damned if we do, damned if we don't situation, sigh.

True. But with the parser pull requests at least there's pretty much no Geany code affected apart from the mappings and the parsers themselves should be reasonably well tested upstream (the unit tests use fuzzing and valgrind). And we'll run into this in the future releases too - we'll just copy the upstream version of all parsers and I think none of us will have enough time to manually test every single filetype for various input. So for the parsers we'll have to more or less trust the quality of the upstream code.

**Comment by elextr** [comment]
> So for the parsers we'll have to more or less trust the quality of the upstream code.

Yes, I thought we had already agreed on that, even the chief inspector @b4n :-), all that any scan would look for is obvious PR mistakes, like the PR changing unrelated Geany code (like my question above about changes to ctags/main).

**Comment by Skif-off** [comment]
[scintilla/lexilla/src/Lexilla.cxx#L191](https://github.com/geany/geany/blob/master/scintilla/lexilla/src/Lexilla.cxx#L191):

```c
		&lmFreeBasic,
+		&lmGDScript,
		&lmHaskell,
```
"+" looks like mistake. Or I'm wrong?

**Comment by Davidy22** [comment]
Oh word, that looks like a diff screwed up somewhere. How does that even happen? Making a pr, how did nothing break for me because of this?

**Comment by elextr** [comment]
@Davidy22 nothing broke because a unary plus operator applied to a valid type (arithmetic, pointer etc) returns the original unchanged, and `&lmGDScript` is a pointer, so it was harmless.

But for sure fix it.

**Comment by Davidy22** [comment]
Oh nice, saved(?) by c idiosyncrasy

**Pull request merged**